cpp/src/io/text/data_chunk_source_factories.cpp

```diff
@@ -0,0 +1,282 @@
+    if (!device_reader) {
+      host_reader = std::make_unique<istream_data_chunk_reader>(
+        std::make_unique<std::ifstream>(_filename, std::ifstream::in));
+    }
+  }
+
+  void skip_bytes(std::size_t size) override
+  {
+    if (device_reader) {
+      bytes_read += size;
+    } else {
+      host_reader->skip_bytes(size);
+    }
+  }
+
+  std::unique_ptr<device_data_chunk> get_next_chunk(std::size_t read_size,
+                                                    rmm::cuda_stream_view stream) override
+  {
+    if (device_reader) {
+      auto chunk = rmm::device_uvector<char>(read_size, stream);
+      auto size  = device_reader->read(bytes_read, read_size, chunk.data(), stream);
+      bytes_read += size;
+      return std::make_unique<device_uvector_data_chunk>(std::move(chunk), size, 0);
+    } else {
+      return host_reader->get_next_chunk(read_size, stream);
+    }
+  }
+
+ private:
+  std::string _filename;
+  std::unique_ptr<cudf::io::detail::cufile_input_impl> device_reader;
+  std::unique_ptr<istream_data_chunk_reader> host_reader;
+  size_t bytes_read = 0;
+};
+
+/**
+ * @brief a reader which produces view of device memory which represent a subset of the input device
+ * span
+ *
+ */
+class device_span_data_chunk_reader : public data_chunk_reader {
+ public:
+  device_span_data_chunk_reader(device_span<char const> data) : _data(data) {}
+
+  void skip_bytes(std::size_t read_size) override
+  {
+    if (read_size > _data.size() - _position) { read_size = _data.size() - _position; }
+    _position += read_size;
+  };
+
+  std::unique_ptr<device_data_chunk> get_next_chunk(std::size_t read_size,
+                                                    rmm::cuda_stream_view stream) override
+  {
+    // limit the read size to the number of bytes remaining in the device_span.
+    if (read_size > _data.size() - _position) { read_size = _data.size() - _position; }
+
+    // create a view over the device span
+    auto chunk_span = _data.subspan(_position, read_size);
+
+    // increment position
+    _position += read_size;
+
+    // return the view over device memory so it can be processed.
+    return std::make_unique<device_span_data_chunk>(chunk_span);
+  }
+
+ private:
+  device_span<char const> _data;
+  uint64_t _position = 0;
+};
+
+/**
+ * @brief a file data source which creates an istream_data_chunk_reader
+ *
+ */
+class file_data_chunk_source : public data_chunk_source {
+ public:
+  file_data_chunk_source(std::string filename) : _filename(filename) {}
+  [[nodiscard]] std::unique_ptr<data_chunk_reader> create_reader() const override
+  {
+    static const bool use_gds = std::getenv("MULTIBYTE_SPLIT_USE_GDS") != nullptr;
+    if (use_gds) {
+      return std::make_unique<istream_data_chunk_reader>(
+        std::make_unique<std::ifstream>(_filename, std::ifstream::in));
+    } else {
+      return std::make_unique<gds_data_chunk_reader>(_filename);
+    }
```

Comment:
instead of having gds_data_chunk_reader depend on host_data_chunk_reader, we can throw an exception here if gds_data_chunk_reader fails to instantiate, and then return a host_data_chunk_reader instead. that should simplify the gds_data_chunk_reader implementation.

```suggestion
    static const bool use_gds = std::getenv("MULTIBYTE_SPLIT_USE_GDS") != nullptr;
    if (use_gds) {
      try {
        return std::make_unique<gds_data_chunk_reader>(_filename);
      } catch (...) {
       // log
      }
    }
    return std::make_unique<istream_data_chunk_reader>(
          std::make_unique<std::ifstream>(_filename, std::ifstream::in));
```